cmd/auth/main.go

```diff
@@ -240,19 +242,30 @@ func newService(db *sqlx.DB, tracer opentracing.Tracer, secret string, logger lo
 	return svc
 }
 
-func startHTTPServer(tracer opentracing.Tracer, svc auth.Service, port string, certFile string, keyFile string, logger logger.Logger, errs chan error) {
+func startHTTPServer(ctx context.Context, tracer opentracing.Tracer, svc auth.Service, port string, certFile string, keyFile string, logger logger.Logger, errs chan error) {
 	p := fmt.Sprintf(":%s", port)
+	server := &http.Server{Addr: p, Handler: httpapi.MakeHandler(svc, tracer)}
 	if certFile != "" || keyFile != "" {
 		logger.Info(fmt.Sprintf("Authentication service started using https, cert %s key %s, exposed port %s", certFile, keyFile, port))
-		errs <- http.ListenAndServeTLS(p, certFile, keyFile, httpapi.MakeHandler(svc, tracer))
-		return
+		go func() {
+			errs <- server.ListenAndServeTLS(certFile, keyFile)
+			return
+		}()
+	} else {
```

Comment:
Is this `else` really needed?

Reply:
yes, its required, since the function doesn't return in if condition and need to wait for ctx.Done() channel after the if else ,

 Alternatively we can use switch with case true and default 
 
```
switch certFile != "" || keyFile != "" {
case true : 
	logger.Info(fmt.Sprintf("Authentication service started using https, cert %s key %s, exposed port %s", certFile, keyFile, port))
	go func() {
		errs <- server.ListenAndServeTLS(certFile, keyFile)
		return
	}()
default : 
	logger.Info(fmt.Sprintf("Authentication service started using http, exposed port %s", port))
	go func() {
		errs <- server.ListenAndServe()
		return
	}()
}

```

Reply:
This looks better. Even a switch with no condition is OK:
```go
switch {
	case certFile != "" || keyFile != "":
		logger.Info(fmt.Sprintf("Authentication service started using https, cert %s key %s, exposed port %s", certFile, keyFile, port))
		go func() {
			errs <- server.ListenAndServeTLS(certFile, keyFile)
		}()
	default:
		logger.Info(fmt.Sprintf("Authentication service started using http, exposed port %s", port))
		go func() {
			errs <- server.ListenAndServe()
		}()
	}
	<-ctx.Done()
	if err := server.Shutdown(context.Background()); err != nil {
		errs <- fmt.Errorf("server shutdown failed:%v", err)
	}
```
Please note that `return` is removed since it is the last statement in the function. Also, notice `fmt.Errorf("server shutdown failed:%v", err)` because errors should not be capitalized and there is no need for `%+v`, but just `%v`.